src/execution_plan/optimizations/traverse_order_utils.c

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright 2018-2020 Redis Labs Ltd. and Contributors
+ *
+ * This file is available under the Redis Labs Source Available License Agreement
+ */
+
+#include "RG.h"
+#include "../../util/arr.h"
+#include "../../util/strcmp.h"
+#include "traverse_order_utils.h"
+
+//------------------------------------------------------------------------------
+// Scoring functions
+//------------------------------------------------------------------------------
+
+// score expression based on its source and destination labels
+// score = src label count + dest label count
+int TraverseOrder_LabelsScore(AlgebraicExpression *exp, const QueryGraph *qg) {
+	ASSERT(exp != NULL);
+	ASSERT(qg  != NULL);
+
+	int         score       =  0;
+	const char  *src        =  AlgebraicExpression_Source(exp);
+	const char  *dest       =  AlgebraicExpression_Destination(exp);
+	QGNode      *src_node   =  QueryGraph_GetNodeByAlias(qg, src);
+	QGNode      *dest_node  =  QueryGraph_GetNodeByAlias(qg, dest);
+
+	score += QGNode_LabelCount(src_node);
+	score += QGNode_LabelCount(dest_node);
+	return score;
+}
+
+// score expression based on the existence of filters applied to entities
+// in the expression, adding additional score for independent filtered entities
+int TraverseOrder_FilterExistenceScore(AlgebraicExpression *exp,
+									   rax *filtered_entities) {
+	ASSERT(exp != NULL);
+
+	if(!filtered_entities) return 0;
+
+	int          score          =  0;
+	void         *frequency     =  NULL;
+	const  char  *src           =  AlgebraicExpression_Source(exp);
+	const  char  *dest          =  AlgebraicExpression_Destination(exp);
+	const  char  *edge          =  AlgebraicExpression_Edge(exp);
+
+	frequency = raxFind(filtered_entities, (unsigned char *)src, strlen(src));
+	if(frequency != raxNotFound) {
+		score += 2;
+		score += (int64_t)frequency * 2;
+	}
+
+	frequency = raxFind(filtered_entities, (unsigned char *)dest, strlen(dest));
+	if(frequency != raxNotFound) {
+		score += 2;
+		score += (int64_t)frequency * 2;
+	}
+
+	if(edge) {
+		if(raxFind(filtered_entities, (unsigned char *)edge, strlen(edge))
+		   != raxNotFound) {
+			score += 1;
+		}
```

Comment:
Add comment describing logic for scoring edge predicates.

Reply:
👍